--- client.py	2017-09-07 16:30:19.177633538 +0000
+++ client.py.final	2017-09-07 17:04:43.248499027 +0000
@@ -626,15 +626,14 @@
         :raises: WSManInvalidResponse when receiving invalid response
         """
         if wait_for_idrac:
-            self.wait_until_idrac_is_ready(self._ready_retries,
-                                           self._ready_retry_delay)
+            self.wait_until_idrac_is_ready()
 
         return super(WSManClient, self).enumerate(resource_uri, optimization,
                                                   max_elems, auto_pull,
                                                   filter_query, filter_dialect)
 
     def invoke(self, resource_uri, method, selectors=None, properties=None,
-               expected_return_value=None, wait_for_idrac=True):
+               expected_return_value=None, check_return_value=True, wait_for_idrac=True):
         """Invokes a remote WS-Man method
 
         :param resource_uri: URI of the resource
@@ -656,8 +655,7 @@
         :raises: DRACUnexpectedReturnValue on return value mismatch
         """
         if wait_for_idrac:
-            self.wait_until_idrac_is_ready(self._ready_retries,
-                                           self._ready_retry_delay)
+            self.wait_until_idrac_is_ready()
 
         if selectors is None:
             selectors = {}
@@ -668,17 +666,18 @@
         resp = super(WSManClient, self).invoke(resource_uri, method, selectors,
                                                properties)
 
-        return_value = utils.find_xml(resp, 'ReturnValue', resource_uri).text
-        if return_value == utils.RET_ERROR:
-            message_elems = utils.find_xml(resp, 'Message', resource_uri, True)
-            messages = [message_elem.text for message_elem in message_elems]
-            raise exceptions.DRACOperationFailed(drac_messages=messages)
-
-        if (expected_return_value is not None and
-                return_value != expected_return_value):
-            raise exceptions.DRACUnexpectedReturnValue(
-                expected_return_value=expected_return_value,
-                actual_return_value=return_value)
+        if check_return_value:
+            return_value = utils.find_xml(resp, 'ReturnValue', resource_uri).text
+            if return_value == utils.RET_ERROR:
+                message_elems = utils.find_xml(resp, 'Message', resource_uri, True)
+                messages = [message_elem.text for message_elem in message_elems]
+                raise exceptions.DRACOperationFailed(drac_messages=messages)
+
+            if (expected_return_value is not None and
+                    return_value != expected_return_value):
+                raise exceptions.DRACUnexpectedReturnValue(
+                    expected_return_value=expected_return_value,
+                    actual_return_value=return_value)
 
         return resp
 
