--- /usr/lib/python2.7/site-packages/dracclient/client.py	2019-02-15 20:14:23.863509901 +0000
+++ client.py	2019-02-15 20:14:16.685378567 +0000
@@ -272,7 +272,7 @@
         state_reached = self._wait_for_host_state(
             self.client.host,
             alive=False,
-            ping_count=3,
+            ping_count=2,
             retries=24)
 
         if not state_reached:
@@ -785,7 +785,7 @@
         return self._raid_mgmt.delete_virtual_disk(virtual_disk)
 
     def commit_pending_raid_changes(self, raid_controller, reboot=False,
-                                    start_time='TIME_NOW'):
+                                    start_time='TIME_NOW', realtime=False):
         """Applies all pending changes on a RAID controller
 
          ...by creating a config job.
@@ -811,7 +811,8 @@
             cim_name='DCIM:RAIDService',
             target=raid_controller,
             reboot=reboot,
-            start_time=start_time)
+            start_time=start_time,
+            realtime=realtime)
 
     def abandon_pending_raid_changes(self, raid_controller):
         """Deletes all pending changes on a RAID controller
