--- /usr/lib/python2.7/site-packages/dracclient/client.py	2018-10-24 22:11:02.639033660 +0000
+++ client.py	2018-10-24 21:43:54.734033660 +0000
@@ -569,7 +569,8 @@
         """
         return self._raid_mgmt.delete_virtual_disk(virtual_disk)
 
-    def commit_pending_raid_changes(self, raid_controller, reboot=False):
+    def commit_pending_raid_changes(self, raid_controller, reboot=False,
+                                    realtime=False):
         """Applies all pending changes on a RAID controller
 
          ...by creating a config job.
@@ -587,7 +588,8 @@
         return self._job_mgmt.create_config_job(
             resource_uri=uris.DCIM_RAIDService,
             cim_creation_class_name='DCIM_RAIDService',
-            cim_name='DCIM:RAIDService', target=raid_controller, reboot=reboot)
+            cim_name='DCIM:RAIDService', target=raid_controller, reboot=reboot,
+            realtime=realtime)
 
     def abandon_pending_raid_changes(self, raid_controller):
         """Deletes all pending changes on a RAID controller
