--- /usr/lib/python2.7/site-packages/dracclient/client.py	2018-10-31 21:20:56.202000000 +0000
+++ client.py	2018-10-31 21:20:32.312000000 +0000
@@ -687,7 +687,7 @@
         return self._raid_mgmt.delete_virtual_disk(virtual_disk)
 
     def commit_pending_raid_changes(self, raid_controller, reboot=False,
-                                    start_time='TIME_NOW'):
+                                    start_time='TIME_NOW', realtime=False):
         """Applies all pending changes on a RAID controller
 
          ...by creating a config job.
@@ -713,7 +713,8 @@
             cim_name='DCIM:RAIDService',
             target=raid_controller,
             reboot=reboot,
-            start_time=start_time)
+            start_time=start_time,
+            realtime=realtime)
 
     def abandon_pending_raid_changes(self, raid_controller):
         """Deletes all pending changes on a RAID controller
