--- /usr/lib/python2.7/site-packages/dracclient/constants.py	2018-02-19 12:01:19.000000000 +0000
+++ constants.py	2018-08-16 19:21:12.446870000 +0000
@@ -12,7 +12,7 @@
 #    under the License.
 
 # iDRAC is ready retry constants
-DEFAULT_IDRAC_IS_READY_RETRIES = 48
+DEFAULT_IDRAC_IS_READY_RETRIES = 96
 DEFAULT_IDRAC_IS_READY_RETRY_DELAY_SEC = 10
 
 # Web Services Management (WS-Management and WS-Man) SSL retry on error
