--- raid.py	2021-09-16 14:39:49.035032177 -0500
+++ raid.new	2021-09-16 14:38:50.658107416 -0500
@@ -911,6 +911,15 @@
         RAID to JBOD or vice versa.  It does this by only converting the
         disks that are not already in the correct state.
 
+        When converting H755 RAID controller physical disks to non-RAID mode,
+        RAID-0 virtual disks get created for each physical disk and disks moved
+        to 'Online' state.
+
+        This is different from other controllers supporting non-RAID conversion
+        and takes up physical disks that cannot be used for user intended RAID
+        configuration later. H755 RAID controllers are excluded from converting
+        to non-RAID mode leaving disks in 'Ready' state.
+
         :param mode: constants.RaidStatus enumeration that indicates the mode
                      to change the disks to.
         :param controllers_to_physical_disk_ids: Dictionary of controllers and
@@ -939,7 +948,9 @@
         physical_disks = self.list_physical_disks()
 
         raid = constants.RaidStatus.raid
+        jbod = constants.RaidStatus.jbod
 
+        all_controllers = self.list_raid_controllers()
         if not controllers_to_physical_disk_ids:
             controllers_to_physical_disk_ids = collections.defaultdict(list)
 
@@ -952,6 +963,25 @@
                         physical_d.controller]
 
                     physical_disk_ids.append(physical_d.id)
+        # Filter out PERC H755 controller as it creates RAID0 virtual disks
+        # when in non-RAID mode
+        controllers_to_results = {}
+
+
+        if mode == jbod:
+            for cntlr in all_controllers:
+                if cntlr.model.startswith("PERC H755") and \
+                 cntlr.id in controllers_to_physical_disk_ids:
+                    LOG.debug("Excluding {} from converting to "
+                              "non-RAID mode".format(cntlr.model))
+                    del controllers_to_physical_disk_ids[cntlr.id]
+                    controllers_to_results[cntlr.id] = \
+                        utils.build_return_dict(
+                            doc=None,
+                            resource_uri=None,
+                            is_commit_required_value=False,
+                            is_reboot_required_value=constants.
+                            RebootRequired.false)
 
         '''Modify controllers_to_physical_disk_ids dict by inspecting desired
         status vs current status of each controller's disks.
