--- /usr/lib/python2.7/site-packages/dracclient/resources/raid.py	2018-03-26 12:41:14.000000000 -0500
+++ /usr/lib/python2.7/site-packages/dracclient/resources/raid.new	2018-09-28 06:52:19.646526000 -0500
@@ -428,10 +428,7 @@
             error_msgs.append("'physical_disks' is not supplied")
 
         # size validation
-        if not size_mb:
-            error_msgs.append("'size_mb' is not supplied")
-        else:
-            utils.validate_integer_value(size_mb, 'size_mb', error_msgs)
+        utils.validate_integer_value(size_mb, 'size_mb', error_msgs)
 
         virtual_disk_prop_names.append('Size')
         virtual_disk_prop_values.append(str(size_mb))
