--- inspect_orig.py	2019-11-08 15:00:52.263876359 +0000
+++ inspect_final.py	2019-11-08 15:03:05.941596489 +0000
@@ -23,6 +23,7 @@
 from ironic.common import exception
 from ironic.common.i18n import _
 from ironic.common import states
+from ironic.common import utils
 from ironic.drivers import base
 from ironic.drivers.modules.drac import common as drac_common
 from ironic import objects
@@ -89,6 +90,21 @@
                     [self._calculate_cpus(cpu) for cpu in cpus])
                 properties['cpu_arch'] = 'x86_64' if cpus[0].arch64 else 'x86'
 
+            bios_settings = client.list_bios_settings()
+            current_capabilities = node.properties.get('capabilities', '')
+            new_capabilities = {
+                'boot_mode': bios_settings["BootMode"].current_value.lower()}
+            capabilties = utils.get_updated_capabilities(current_capabilities,
+                                                         new_capabilities)
+            properties['capabilities'] = capabilties
+            bios_settings = client.list_bios_settings()
+            current_capabilities = node.properties.get('capabilities', '')
+            new_capabilities = {
+                'boot_mode': bios_settings["BootMode"].current_value.lower()}
+            capabilties = utils.get_updated_capabilities(current_capabilities,
+                                                         new_capabilities)
+            properties['capabilities'] = capabilties
+
             virtual_disks = client.list_virtual_disks()
             root_disk = self._guess_root_disk(virtual_disks)
             if root_disk:
