--- /httpboot/inspector.ipxe	2017-08-16 06:52:50.055384180 -0500
+++ inspector.ipxe	2017-08-16 06:56:34.581967489 -0500
@@ -2,6 +2,6 @@
 
 :retry_boot
 imgfree
-kernel --timeout 60000 http://192.168.120.112:8088/agent.kernel ipa-inspection-callback-url=http://192.168.120.112:5050/v1/continue ipa-inspection-collectors=default,extra-hardware,logs systemd.journald.forward_to_console=yes BOOTIF=${mac} ipa-debug=1 ipa-inspection-dhcp-all-interfaces=1 ipa-collect-lldp=1 initrd=agent.ramdisk || goto retry_boot
+kernel --timeout 60000 http://192.168.120.112:8088/agent.kernel ipa-inspection-callback-url=http://192.168.120.112:5050/v1/continue ipa-inspection-collectors=default,extra-hardware,logs systemd.journald.forward_to_console=yes BOOTIF=${mac} ipa-debug=1 ipa-inspection-dhcp-all-interfaces=1 ipa-collect-lldp=1 initrd=agent.ramdisk net.ifnames=0 biosdevname=0 || goto retry_boot
 initrd --timeout 60000 http://192.168.120.112:8088/agent.ramdisk || goto retry_boot
 boot
