--- ironic.conf	2018-07-12 14:04:55.353000000 -0500
+++ ironic.conf.new	2018-07-12 14:04:23.947000000 -0500
@@ -1294,7 +1294,7 @@
 
 # Interval between syncing the node power state to the
 # database, in seconds. (integer value)
-#sync_power_state_interval = 60
+sync_power_state_interval = 70
 
 # Interval between checks of provision timeouts, in seconds.
 # (integer value)
@@ -1327,7 +1327,7 @@
 #periodic_max_workers = 8
 
 # Number of attempts to grab a node lock. (integer value)
-#node_locked_retry_attempts = 3
+node_locked_retry_attempts = 25
 
 # Seconds to sleep between node lock attempts. (integer value)
 #node_locked_retry_interval = 1
@@ -1393,7 +1393,7 @@
 # doing the cleaning. If the timeout is reached the node will
 # be put in the "clean failed" provision state. Set to 0 to
 # disable timeout. (integer value)
-#clean_callback_timeout = 1800
+clean_callback_timeout = 5400
 
 # Timeout (seconds) to wait for a callback from the rescue
 # ramdisk. If the timeout is reached the node will be put in
