--- /etc/ironic/ironic.conf	2017-11-13 18:46:02.757143022 +0000
+++ /root/ironic.conf.new	2017-11-13 18:47:09.636378764 +0000
@@ -655,7 +655,7 @@
 
 # Interval between syncing the node power state to the
 # database, in seconds. (integer value)
-#sync_power_state_interval = 60
+sync_power_state_interval = 7200
 
 # Interval between checks of provision timeouts, in seconds.
 # (integer value)
@@ -683,7 +683,7 @@
 #periodic_max_workers = 8
 
 # Number of attempts to grab a node lock. (integer value)
-#node_locked_retry_attempts = 3
+node_locked_retry_attempts = 15
 
 # Seconds to sleep between node lock attempts. (integer value)
 #node_locked_retry_interval = 1
