--- /usr/lib/python2.7/site-packages/dracclient/resources/lifecycle_controller.py	2017-03-01 16:12:40.000000000 +0000
+++ lifecycle_controller.py	2017-08-18 11:34:54.689100174 +0000
@@ -35,10 +35,7 @@
                  interface
         """
 
-        filter_query = ('select LifecycleControllerVersion '
-                        'from DCIM_SystemView')
-        doc = self.client.enumerate(uris.DCIM_SystemView,
-                                    filter_query=filter_query)
+        doc = self.client.enumerate(uris.DCIM_SystemView, wait_for_idrac=False)
         lc_version_str = utils.find_xml(doc, 'LifecycleControllerVersion',
                                         uris.DCIM_SystemView).text
 
