--- /usr/lib/python2.7/site-packages/dracclient/resources/uris.py	2018-02-09 20:36:04.487472252 +0000
+++ uris.py	2018-02-09 20:35:35.961947034 +0000
@@ -76,6 +76,9 @@
 DCIM_PhysicalDiskView = ('http://schemas.dell.com/wbem/wscim/1/cim-schema/2/'
                          'DCIM_PhysicalDiskView')
 
+DCIM_PCIeSSDView = ('http://schemas.dell.com/wbem/wscim/1/cim-schema/2/'
+                    'DCIM_PCIeSSDView')
+
 DCIM_RAIDService = ('http://schemas.dell.com/wbem/wscim/1/cim-schema/2/'
                     'DCIM_RAIDService')
 
