--- vendor_passthru.py	2016-10-06 16:10:34.000000000 +0000
+++ vendor_passthru.py.new	2016-11-03 13:49:03.940733759 +0000
@@ -56,9 +56,7 @@
         """
         bios_attrs = {}
         for name, bios_attr in drac_bios.get_config(task.node).items():
-            # NOTE(ifarkas): call from python-dracclient returns list of
-            #                namedtuples, converting it to dict here.
-            bios_attrs[name] = bios_attr._asdict()
+            bios_attrs[name] = bios_attr.__dict__
 
         return bios_attrs
 
