--- 10-ironic_wsgi.conf	2018-08-14 04:51:39.895122000 -0500
+++ 10-ironic_wsgi.conf.new	2018-08-14 04:52:12.427122000 -0500
@@ -23,7 +23,7 @@
   CustomLog "/var/log/httpd/ironic_wsgi_access.log" combined 
   SetEnvIf X-Forwarded-Proto https HTTPS=1
   WSGIApplicationGroup %{GLOBAL}
-  WSGIDaemonProcess ironic display-name=ironic_wsgi group=ironic processes=4 threads=15 user=ironic
+  WSGIDaemonProcess ironic display-name=ironic_wsgi group=ironic processes=4 threads=30 user=ironic
   WSGIProcessGroup ironic
   WSGIScriptAlias / "/var/www/cgi-bin/ironic/app"
 </VirtualHost>
