--- wsman.py	2019-07-12 06:05:33.739791639 -0500
+++ wsman.new	2019-07-12 06:28:59.132754435 -0500
@@ -25,6 +25,8 @@
 
 LOG = logging.getLogger(__name__)
 
+MAX_ELEMENTS = 384
+
 NS_SOAP_ENV = 'http://www.w3.org/2003/05/soap-envelope'
 NS_WS_ADDR = 'http://schemas.xmlsoap.org/ws/2004/08/addressing'
 NS_WS_ADDR_ANONYM_ROLE = ('http://schemas.xmlsoap.org/ws/2004/08/addressing/'
@@ -135,7 +137,7 @@
         else:
             return resp
 
-    def enumerate(self, resource_uri, optimization=True, max_elems=100,
+    def enumerate(self, resource_uri, optimization=True, max_elems=MAX_ELEMENTS,
                   auto_pull=True, filter_query=None, filter_dialect='cql'):
         """Executes enumerate operation over WSMan.
 
@@ -200,7 +202,7 @@
         else:
             return resp_xml
 
-    def pull(self, resource_uri, context, max_elems=100):
+    def pull(self, resource_uri, context, max_elems=MAX_ELEMENTS):
         """Executes pull operation over WSMan.
 
         :param resource_uri: URI of resource to pull
@@ -291,7 +293,7 @@
     """Payload generation for WSMan enumerate operation."""
 
     def __init__(self, endpoint, resource_uri, optimization=True,
-                 max_elems=100, filter_query=None, filter_dialect=None):
+                 max_elems=MAX_ELEMENTS, filter_query=None, filter_dialect=None):
         self.endpoint = endpoint
         self.resource_uri = resource_uri
         self.filter_dialect = None
@@ -351,7 +353,7 @@
 class _PullPayload(_Payload):
     """Payload generation for WSMan pull operation."""
 
-    def __init__(self, endpoint, resource_uri, context, max_elems=100):
+    def __init__(self, endpoint, resource_uri, context, max_elems=MAX_ELEMENTS):
         self.endpoint = endpoint
         self.resource_uri = resource_uri
         self.context = context
